lib/Dialects/VectorExt/Transform/VectorMaskingUtils.cpp

```diff
@@ -0,0 +1,138 @@
+}
+
+/// Iterate over all the definitions produced by the operation range `opRange`
+/// and gather those with users outside the range.
+static void
+getDefsWithExternalUses(llvm::iterator_range<Block::iterator> opRange,
+                        SmallVectorImpl<Value> &defsWithExternalUses,
+                        SmallVectorImpl<Type> &defTypes) {
+  SmallPtrSet<Operation *, 8> rangeSet;
+  for (Operation &op : opRange)
+    rangeSet.insert(&op);
+
+  for (Operation &op : llvm::make_early_inc_range(opRange))
+    for (Value res : op.getResults())
+      for (Operation *user : res.getUsers())
+        if (rangeSet.count(user) == 0) {
+          defsWithExternalUses.push_back(res);
+          defTypes.push_back(res.getType());
+          break;
+        }
+}
+
+/// Introduce a vector.predicate op that encloses the loop body of the tiled
+/// loop. The predicate is computed based on the provided vectorization factor.
+static void createVectorPredicateOp(TiledLoopOp loopOp, int64_t vecFactor) {
+  // Compute the range of operations that will be moved within vector.predicate
+  // and the definitions within the range with external users.
+  Block &bodyBlock = loopOp.getLoopBody().front();
+  auto opsToMove =
+      llvm::make_range(bodyBlock.begin(), Block::iterator(bodyBlock.back()));
+
+  SmallVector<Value, 8> defsWithExternalUses;
+  SmallVector<Type, 8> defTypes;
+  getDefsWithExternalUses(opsToMove, defsWithExternalUses, defTypes);
+
+  // Initialize a builder to insert vector.predicate at the beginning of the
+  // loop body.
+  Location loc = loopOp.getLoc();
+  MLIRContext *context = loopOp.getContext();
+  OpBuilder builder(loopOp);
+  builder.setInsertionPointToStart(&bodyBlock);
+
+  // Generate the predicate to used by the vector.predicate operation:
+  //   %min = affine.min affine_map<(d0)[s0] -> (vecFactor, -d0 + s0)>(%iv)[%ub]
+  //   %vec_pred = vector.create_mask %min : vector<8xi1>
+  // TODO: This is probably not the most efficient way to generate the vector
+  // predicate. Consider using the vector IV.
+  AffineExpr i, j;
+  bindDims(context, i, j);
+  SmallVector<AffineMap, 4> maps = AffineMap::inferFromExprList(
+      ExprList{{builder.getAffineConstantExpr(vecFactor), i - j}});
+  auto minOp = builder.create<AffineMinOp>(
+      loc, loopOp.step()[0].getType(), maps[0],
+      ValueRange{loopOp.upperBound()[0], loopOp.getInductionVars()[0]});
+
+  auto maskType = VectorType::get({vecFactor}, builder.getI1Type());
+  auto predicate =
+      builder.create<CreateMaskOp>(loc, maskType, ValueRange{minOp});
+
+  // Generate the vector.predicate operation and move the range [`startOp`,
+  // `endOp`] of operations within its truePredicateRegion. We have to rewire
+  // the def-use chain for those definitions within the range that have external
+  // uses. Those uses are rewired to the results of the vector.predicate.
+  auto vecPredOp = builder.create<PredicateOp>(loc, defTypes, predicate);
```

Comment:
Some of these may look quite different with a PatternRewriter, can this be isolated in a helper that takes the builder so it becomes easier to both 1) distinguish the mechanical stuff and 2) evolve as needed ?

Reply:
Let me know what you think about the new version of the code. I changed the utility so that it returns this newly created op. Then, an external rewriter could use it to replace another op, for example. Would that be enough?

Reply:
yes, looks great, the idea was to avoid cloning the builder internally which would create a mess when mixed with PatterRewriter.